fs_overlay/etc/services.d/30-dynamic-env/run

```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
+dir=/var/lib/https-portal/dynamic-env
+mkdir -p $dir
+
+# Remember when we started (tiny race condition window)
+touch $dir.stamp # Outside of watched dir!
+
+# Wait for a modification to an all-uppercase file (ignore temp/swp files)
+inotifywait -qm -e close_write,move,delete $dir | while read line
+do
+	if echo "$line" | grep -q ' [_A-Z]*$'
+	then
+		echo "[dynamic-env] All-uppercase file modified in $dir" >&2
+
+		echo -n "[dynamic-env] Wait for activity to cease..." >&2
+		while inotifywait -qq -t 1 -e close_write,move,delete $dir
+		do
+			echo -n . >&2
+		done
+		echo " -- ceased" >&2
+
+		# Apply new configuration; but only, if anything actually
+		# changed (top-level "while" loop will continue getting
+		# potentially old events)
+		if find $dir -newer $dir.stamp | grep -q .
+		then
+			# Remember when we started (tiny race condition window)
+			touch $dir.stamp # Outside of watched dir!
+
+			echo "[dynamic-env] File changed since last run: Applying new configuration..." >&2
+			/usr/bin/justc-envdir $dir /bin/reconfig
+			echo "[dynamic-env] Configuration applied" >&2
+		else
+			echo "[dynamic-env] No file changed since last run -- suppressing" >&2
+		fi
+	fi
+done
```

Comment:
Part of me want to rewrite this piece of code with Ruby because I hate maintaining shell scripts. 
However I'm afraid memory consumption might go up quite a bit, since this process will be in memory all the time.

Reply:
I also do not think a shell script is ideal here. While I believe it is solid in practice, it is somewhat awkward as a piece of code.

* The watch/debounce logic has race conditions. However, it should not affect anything in practice: `cp`, `vi`, `git pull` and probably many more "standard" operations anyone will use here should be covered.
* There are two layers of notifications (the debounce loop will read events that the outside loop is getting as well, having to suppress them later); a single event loop would simplify things.
* I am not happy with the use of a timestamp file, using an internal time variable to compare against `stat.st_mtime` would be better.

I considered rewriting it in Python (my weapon of choice; haven't done a lot of Ruby in the past few years), but did not want to introduce yet another tool (even though I just noticed `acme_tiny` already uses it). However, I think the code would not become smaller and more readable, as implementing the functions of `inotifywait` and `find` probably would create much more boilerplate code.

If you prefer, I can give a try in Python.

Reply:
Nah, don't worry about it. I will probably rewrite it in Ruby if I find it necessary.
The CertManager has `with_lock` block which prevents two processes obtaining certificates at the same time. If we make the reconfig idempotent and fast, a bit of bouncing should be fine.

Reply:
> However I'm afraid memory consumption might go up quite a bit, since this process will be in memory all the time.

I guess the set of active pages for such a process would be rather small. And there would be page sharing with other Ruby processes, notably CertManager. So I do not think memory should be a show-stopper.

Reply:
CertManager only runs when doing `setup` or `reconfig`. If I write this whole piece with Ruby, it will stay in memory forever. And Ruby interpreter is kinda notorious for its memory consumption ...

That's what concerns me a little. 

Of course, we can find some other ways such as having a process invoking a Ruby script when it detects changes. I will explore later. But for now, the Shell script is fine.